driver/transform-dsl.md

```diff
@@ -0,0 +1,549 @@
+
+```json
+{
+    "x": null
+}
+```
+
+These are the only constraints that were defined for this transform, but
+we still have an "offset" field left. At this point `Check` will fail
+because "offset" field was not handled by a transform. This usually means
+that AST shape was changed unexpectedly, or developer forgot to describe
+this field. Both cases indicate an error that must be addressed
+(see Completeness section for more details).
+
+But we can define our transform as partial and it will automatically
+store all unhandled field to node state, which is useful if e want to
+change only specific fields:
+
+```golang
+Part("other", Obj{ ... })
+```
+
+In this case, `Check` will succeed and produce the following node state:
+
+```json
+{
+    "x": null,
+    "other": {"offset": 5}
+}
+```
+
+Now, we will reconstruct an AST tree by using `Construct` method and
+the state for this node that we just received:
+
+```
+newAST, _ := op.Construct(s, nil)
+```
+
+In this case process runs in reverse - each operation uses state to
+reconstruct AST nodes based on defined constraints.
+
+First, an `Obj` operation will construct an empty AST node:
+
+```json
+{}
+```
+
+Then, it will start to populate fields:
+
+```golang
+"type": String("Ident"),
+```
+
+For this field, everything is strictly defined and no state is required
+to reconstruct the field - it will simply assign a "type" to "Ident".
+
+```json
+{
+    "type": "Ident"
+}
+```
+
+For the "name" field the state is required to reconstruct it:
+
+```golang
+"name": Var("x"),
+```
+
+At this point, state variable "x" will be used to reconstruct the field
+value (assign it to null).
+
+Remember that variables can store any node type: it might have been
+a string, an array or another AST node.
+
+```json
+{
+    "type": "Ident",
+    "name": null
+}
+```
+
+Since there is no more constraints defined, `Part` transform will use
+it's state stored in "other" variable to populate all unhandled fields:
+
+```json
+{
+    "type": "Ident",
+    "name": null,
+    "offset": 5
+}
+```
+
+As you can see, an operation can be used to describe a tree shape and
+reconstruct it back, assuming that all placeholder variables used in
+an operation are defined in node state.
+
+### Mappings
+
+Since mapping defines two operations (source and target), we can change
+our code to use one operation for checking the tree and building
+the state and a different operation for constructing it:
+
+```golang
+A := Part("other", Obj{
+    "type": String("Ident"),
+    "name": Var("x"),
+})
+
+B := Part("other", Obj{
+    "type": String("go:Ident"),
+    "token": Var("x"),
+})
+
+s := NewState()
+
+// verify an AST with operation A, populate state
+if ok, _ := A.Check(s, ast); ok {
+    // construct a new AST using the state and operation B
+    newAST, _ := B.Construct(s, nil)
+}
+```
+
+As long as all variable defined in A exactly match all variables defined
+in B, `Check`-`Construct` routine will work the same way as described
```

Comment:
**the** `Check` - `Construct` routine will work **in** the same way as described above (or in **the** example section above).